// ngx-tui-calendar/esm2015/lib/Models/Events.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function AfterRenderScheduleEvent() { }
/** @type {?} */
AfterRenderScheduleEvent.prototype.schedule;
/**
 * @record
 */
export function BeforeCreateScheduleEvent() { }
/** @type {?} */
BeforeCreateScheduleEvent.prototype.isAllDay;
/** @type {?} */
BeforeCreateScheduleEvent.prototype.start;
/** @type {?} */
BeforeCreateScheduleEvent.prototype.end;
/** @type {?} */
BeforeCreateScheduleEvent.prototype.guide;
/** @type {?} */
BeforeCreateScheduleEvent.prototype.triggerEventName;
/**
 * @record
 */
export function BeforeDeleteScheduleEvent() { }
/** @type {?} */
BeforeDeleteScheduleEvent.prototype.schedule;
/**
 * @record
 */
export function BeforeUpdateScheduleEvent() { }
/** @type {?} */
BeforeUpdateScheduleEvent.prototype.schedule;
/** @type {?} */
BeforeUpdateScheduleEvent.prototype.start;
/** @type {?} */
BeforeUpdateScheduleEvent.prototype.end;
/**
 * @record
 */
export function ClickDaynameEvent() { }
/** @type {?} */
ClickDaynameEvent.prototype.date;
/**
 * @record
 */
export function ClickScheduleEvent() { }
/** @type {?} */
ClickScheduleEvent.prototype.schedule;
/** @type {?} */
ClickScheduleEvent.prototype.event;
