// ngx-tui-calendar/esm2015/lib/ngx-tui-calendar-defaults.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function TuiCalendarOptions() { }
/** @type {?|undefined} */
TuiCalendarOptions.prototype.defaultView;
/** @type {?|undefined} */
TuiCalendarOptions.prototype.taskView;
/** @type {?|undefined} */
TuiCalendarOptions.prototype.scheduleView;
/** @type {?|undefined} */
TuiCalendarOptions.prototype.template;
/** @type {?|undefined} */
TuiCalendarOptions.prototype.month;
/** @type {?|undefined} */
TuiCalendarOptions.prototype.week;
/** @type {?|undefined} */
TuiCalendarOptions.prototype.disableDblClick;
export class TuiCalendarDefaults {
    constructor() {
        this.defaultView = 'week';
        this.taskView = false;
        this.useCreationPopup = false;
        this.useDetailPopup = false;
        this.scheduleView = true;
        this.disableDblClick = false;
        this.week = {
            daynames: ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
            startDayOfWeek: 1,
            narrowWeekend: false,
        };
    }
}
if (false) {
    /** @type {?} */
    TuiCalendarDefaults.prototype.defaultView;
    /** @type {?} */
    TuiCalendarDefaults.prototype.taskView;
    /** @type {?} */
    TuiCalendarDefaults.prototype.useCreationPopup;
    /** @type {?} */
    TuiCalendarDefaults.prototype.useDetailPopup;
    /** @type {?} */
    TuiCalendarDefaults.prototype.scheduleView;
    /** @type {?} */
    TuiCalendarDefaults.prototype.disableDblClick;
    /** @type {?} */
    TuiCalendarDefaults.prototype.week;
}
