// ngx-tui-calendar/esm2015/lib/ngx-tui-calendar.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule, InjectionToken } from '@angular/core';
import { NgxTuiCalendarComponent } from './ngx-tui-calendar.component';
import { TuiCalendarDefaults } from './ngx-tui-calendar-defaults.service';
/** @type {?} */
export const USER_DEFAULTS = new InjectionToken('tuiCalendar defaults');
/**
 * @param {?} userDefaults
 * @return {?}
 */
export function defaultsFactory(userDefaults) {
    /** @type {?} */
    const defaults = new TuiCalendarDefaults();
    Object.assign(defaults, userDefaults);
    return defaults;
}
export class NgxTuiCalendarModule {
    /**
     * @param {?=} userDefaults
     * @return {?}
     */
    static forRoot(userDefaults = {}) {
        return {
            ngModule: NgxTuiCalendarModule,
            providers: [
                {
                    provide: USER_DEFAULTS,
                    useValue: userDefaults
                },
                {
                    provide: TuiCalendarDefaults,
                    useFactory: defaultsFactory,
                    deps: [USER_DEFAULTS]
                }
            ]
        };
    }
}
NgxTuiCalendarModule.decorators = [
    { type: NgModule, args: [{
                declarations: [NgxTuiCalendarComponent],
                exports: [NgxTuiCalendarComponent],
                providers: [{ provide: TuiCalendarDefaults, useClass: TuiCalendarDefaults }]
            },] },
];
